### PR TITLE
vncserver: Support pointer event by uinput

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
Add pointer support to vncserver to make it possible to run interactive GUI application by VNC
## Impact
New feature
## Testing
Custom boards

### CODE REVIEW

File: drivers/video/vnc/vnc_server.c [hunk 1]
```diff
@@ -284,6 +285,19 @@ int vnc_server(int argc, FAR char *argv[])
   nxsem_set_protocol(&session->vsyncsem, SEM_PRIO_NONE);
 #endif
 
+#ifdef CONFIG_VNCSERVER_MOUSE
+  ret = open(CONFIG_VNCSERVER_MOUSE_DEVNAME, O_WRONLY);
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_server.c:289]:
```suggestion
  ret = nx_open(CONFIG_VNCSERVER_MOUSE_DEVNAME, O_WRONLY);
```

File: drivers/video/vnc/vnc_fbdev.c [hunk 2]
```diff
@@ -470,6 +472,42 @@ static int up_waitforsync(FAR struct fb_vtable_s *vtable)
...
+                              TOUCH_POS_VALID;
+    }
+  else
+    {
+      sample.point[0].flags = TOUCH_UP |
+                              TOUCH_ID_VALID |
+                              TOUCH_POS_VALID;
+    }
+
+  return write(session->fd_mouse, (void *)&sample, sizeof(sample));
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_fbdev.c:506]:
```suggestion
  return nx_write(session->fd_mouse, (void *)&sample, sizeof(sample));
```

File: drivers/video/vnc/vnc_fbdev.c [hunk 3]
```diff
@@ -470,6 +472,42 @@ static int up_waitforsync(FAR struct fb_vtable_s *vtable)
...
+  FAR struct vnc_session_s *session = arg;
+  struct touch_sample_s sample;
+
+  DEBUGASSERT(session);
+
+  sample.npoints = 1;
+  sample.point[0].x = x;
+  sample.point[0].y = y;
+
+  if (buttons & MOUSE_BUTTON_1)
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_fbdev.c:493]: Optional
```suggestion
  if ((buttons & MOUSE_BUTTON_1) != 0)
```

File: drivers/video/vnc/vnc_fbdev.c [hunk 4]
```diff
@@ -470,6 +472,42 @@ static int up_waitforsync(FAR struct fb_vtable_s *vtable)
...
+                              TOUCH_POS_VALID;
+    }
+  else
+    {
+      sample.point[0].flags = TOUCH_UP |
+                              TOUCH_ID_VALID |
+                              TOUCH_POS_VALID;
+    }
+
+  return nx_write(session->fd_mouse, (void *)&sample, sizeof(sample));
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:506]: why not implement and register a mouse driver directly?

Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c:506]: This causes VNC mouse driver's private data and VNC session to be coupled together.
And write to uinput is more simple (it's logic is very similar to uinput).

Comment (xiaoxiang781216) [reply drivers/video/vnc/vnc_fbdev.c:506]: But touch screen driver support the upper/lower half in the recent update, which mean writing a touch screen driver isn't complex than send the event through uinput.

Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c]: OK, a touchscreen based driver implemented.

File: drivers/video/vnc/vnc_server.c [hunk 5]
```diff
@@ -284,6 +285,19 @@ int vnc_server(int argc, FAR char *argv[])
   nxsem_set_protocol(&session->vsyncsem, SEM_PRIO_NONE);
 #endif
 
+#ifdef CONFIG_VNCSERVER_MOUSE
+  ret = nx_open(CONFIG_VNCSERVER_MOUSE_DEVNAME, O_WRONLY);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_server.c:289]: revert the change in this file

File: drivers/video/vnc/vnc_touch.c [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: vnc_touch_register
+ ****************************************************************************/
+
+FAR struct touch_lowerhalf_s *vnc_touch_register(FAR const char *devpath)
+{
+  FAR struct touch_lowerhalf_s *vnc_lower;
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:44]: vnc_lower->lower

File: drivers/video/vnc/vnc_touch.c [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    {
+      gerr("Alloc VNC touch driver instance failed.\n");
+      return NULL;
+    }
+
+  ret = touch_register(vnc_lower, devpath, 1);
+
+  if (ret < 0)
+    {
+      gerr("Register VNC touch driver failed.\n");
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:59]: kmm_free(vnc_lower)

File: drivers/video/vnc/vnc_touch.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+  return vnc_lower;
+}
+
+/****************************************************************************
+ * Name: vnc_touch_event
+ ****************************************************************************/
+
+int vnc_touch_event(FAR void *arg, int16_t x,
+                    int16_t y, uint8_t buttons)
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:71]: merge to previous line

File: drivers/video/vnc/vnc_touch.c [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+  return vnc_lower;
+}
+
+/****************************************************************************
+ * Name: vnc_touch_event
+ ****************************************************************************/
+
+int vnc_touch_event(FAR void *arg, int16_t x,
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:70]: int->void

File: drivers/video/vnc/vnc_touch.c [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    {
+      sample.point[0].flags = TOUCH_DOWN |
+                              TOUCH_ID_VALID |
+                              TOUCH_POS_VALID;
+    }
+  else
+    {
+      sample.point[0].flags = TOUCH_UP |
+                              TOUCH_ID_VALID |
+                              TOUCH_POS_VALID;
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:92]: move TOUCH_ID_VALID | TOUCH_POS_VALID; out of if/else to avoid the code dup

File: drivers/video/vnc/vnc_server.h [hunk 11]
```diff
@@ -503,6 +508,37 @@ int vnc_raw(FAR struct vnc_session_s *session, FAR struct fb_area_s *rect);
...
+/****************************************************************************
+ * Name: vnc_touch_event
+ *
+ * Description:
+ *   Report a touch event from vnc client.
+ *
+ ****************************************************************************/
+
+int vnc_touch_event(FAR void *arg, int16_t x,
+                    int16_t y, uint8_t buttons);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_server.h:538]: merge to previous line

File: drivers/video/vnc/Kconfig [hunk 12]
```diff
@@ -139,6 +139,20 @@ config VNCSERVER_KBDENCODE
...
+	select INPUT_TOUCHSCREEN
+	---help---
+		Use uinput based mouse driver
+
+config VNCSERVER_TOUCH_DEVNAME
+	string "Touch input device name"
+	default "/dev/vnctouch"
+	depends on VNCSERVER_TOUCH
+	---help---
+		Device name
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/Kconfig:154]: Touch device name

File: drivers/video/vnc/vnc_fbdev.c [hunk 13]
```diff
@@ -50,6 +50,8 @@
 #include <nuttx/kthread.h>
 #include <nuttx/video/fb.h>
 #include <nuttx/video/vnc.h>
+#include <nuttx/input/touchscreen.h>
+#include <nuttx/input/mouse.h>
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c:54]: `mouse.h` is essential since it provides `MOUSE_BUTTON_1`.

File: drivers/video/vnc/vnc_touch.c (hunk 9 above)
Comment (no1wudi) [reply drivers/video/vnc/vnc_touch.c:70]: Use int here to fit callback prototype `vnc_mouseout_t`

File: drivers/video/vnc/vnc_fbdev.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply drivers/video/vnc/vnc_fbdev.c:54]: Yes, but you don't need reference MOUSE_BUTTON_1 in this file.

File: drivers/video/vnc/vnc_fbdev.c [hunk 14]
```diff
@@ -619,6 +622,22 @@ int up_fbinitialize(int display)
       gerr("ERROR: wait for vnc server start failed: %d\n", ret);
     }
 
+  /* Save the input callout function information in the session structure. */
+
+  session           = g_vnc_sessions[display];
+#ifdef CONFIG_VNCSERVER_TOUCH
+  session->touch    = vnc_touch_register(CONFIG_VNCSERVER_TOUCH_DEVNAME);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:629]: should we call vnc_touch_register before vnc_wait_start? otherwise, the background thread may reference touch before we register the driver.

File: drivers/video/vnc/vnc_server.h [hunk 15]
```diff
@@ -503,6 +508,36 @@ int vnc_raw(FAR struct vnc_session_s *session, FAR struct fb_area_s *rect);
...
+ *
+ * Input Parameters:
+ *   devpath - Device path for this device
+ *
+ * Returned Value:
+ *   Driver instance
+ *
+ ****************************************************************************/
+
+FAR struct touch_lowerhalf_s *vnc_touch_register(FAR const char *devpath);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_server.h:527]: int vnc_touch_register(FAR const char *devpath, FAR struct touch_lowerhalf_s **lower);

File: drivers/video/vnc/vnc_fbdev.c (hunk 14 above)
Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c]: OK, moved before vnc_wait_start

File: drivers/video/vnc/vnc_fbdev.c [hunk 16]
```diff
@@ -610,6 +612,22 @@ int up_fbinitialize(int display)
...
+  /* Save the input callout function information in the session structure. */
+
+  session           = g_vnc_sessions[display];
+#ifdef CONFIG_VNCSERVER_TOUCH
+  ret = vnc_touch_register(CONFIG_VNCSERVER_TOUCH_DEVNAME, &session->touch);
+
+  if (ret < 0)
+    {
+      gerr("Initial vnc touch driver failed.\n");
+      return ERROR;
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:624]: return ret

File: drivers/video/vnc/vnc_fbdev.c [hunk 17]
```diff
@@ -610,6 +612,22 @@ int up_fbinitialize(int display)
       return ret;
     }
 
+  /* Save the input callout function information in the session structure. */
+
+  session           = g_vnc_sessions[display];
+#ifdef CONFIG_VNCSERVER_TOUCH
+  ret = vnc_touch_register(CONFIG_VNCSERVER_TOUCH_DEVNAME, &session->touch);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:619]: we need generate the touch device dynamically to support the multiple instance

File: drivers/video/vnc/vnc_fbdev.c [hunk 18]
```diff
@@ -610,6 +612,22 @@ int up_fbinitialize(int display)
...
+
+  if (ret < 0)
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:621]: sorry, look like the background thread is start at line vnc_start_server, so we need move before vnc_start_server

File: drivers/video/vnc/vnc_server.h [hunk 19]
```diff
@@ -222,6 +223,10 @@ struct vnc_session_s
   vnc_mouseout_t mouseout;     /* Callout when keyboard input is received */
   FAR void *arg;               /* Argument that accompanies the callouts */
 
+#ifdef CONFIG_VNCSERVER_TOUCH
+  FAR struct touch_lowerhalf_s *touch; /* Touch driver instance */
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_server.h:227]: why not embed touch directly in vnc_session_s

File: drivers/video/vnc/Kconfig [hunk 20]
```diff
@@ -139,6 +139,21 @@ config VNCSERVER_KBDENCODE
...
+	bool "Enable touch input"
+	default n
+	select INPUT
+	select INPUT_TOUCHSCREEN
+	---help---
+		Use touchscreen based input driver
+
+config VNCSERVER_TOUCH_DEVNAME
+	string "Touch input device name"
+	default "/dev/vnctouch"
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/Kconfig:152]: look like we need use /dev/inputx like other driver to improve the compatibility.

File: drivers/video/vnc/vnc_fbdev.c [hunk 21]
```diff
@@ -598,9 +601,39 @@ static inline int vnc_wait_start(int display)
...
+
+  ret = vnc_touch_register(devname, session);
+
+  if (ret < 0)
+    {
+      gerr("ERROR: Initial vnc touch driver failed.\n");
+      return ret;
+    }
+
+  session->mouseout = (vnc_mouseout_t)vnc_touch_event;
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:633]: remove the cast

File: drivers/video/vnc/vnc_touch.c [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include "vnc_server.h"
+
+#include <debug.h>
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:27]: remove

File: drivers/video/vnc/vnc_touch.c [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+#include <errno.h>
+
+#include <nuttx/input/touchscreen.h>
+#include <nuttx/input/mouse.h>
+#include <nuttx/kmalloc.h>
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_touch.c:32]: remove

File: drivers/video/vnc/vnc_fbdev.c [hunk 24]
```diff
@@ -22,9 +22,11 @@
  * Included Files
  ****************************************************************************/
 
+#include <limits.h>
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:25]: remove

File: drivers/video/vnc/vnc_fbdev.c [hunk 25]
```diff
@@ -598,9 +600,39 @@ static inline int vnc_wait_start(int display)
...
+      return ret;
+    }
+
+  session->mouseout = vnc_touch_event;
+#endif
+  session->arg      = session;
+
   /* Start the VNC server kernel thread. */
 
   ret = vnc_start_server(display);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:638]: let's unregister touch screen driver in case of fail

Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c:638]: OK, unregister touch on fail now.

**Review by xiaoxiang781216 — APPROVED** [review]
